python/ray/tune/schedulers/pb2.py

```diff
@@ -0,0 +1,341 @@
+            reconstructed.
+        require_attrs (bool): Whether to require time_attr and metric to appear
+            in result for every iteration. If True, error will be raised
+            if these values are not present in trial result.
+        synch (bool): If False, will use asynchronous implementation of
+            PBT. Trial perturbations occur every perturbation_interval for each
+            trial independently. If True, will use synchronous implementation
+            of PBT. Perturbations will occur only after all trials are
+            synced at the same time_attr every perturbation_interval.
+            Defaults to False. See Appendix A.1 here
+            https://arxiv.org/pdf/1711.09846.pdf.
+
+    Example:
+        >>> pb2 = PB2(
+        >>>     time_attr="timesteps_total",
+        >>>     metric="episode_reward_mean",
+        >>>     mode="max",
+        >>>     perturbation_interval=10000,
+        >>>     hyperparam_mutations={
+        >>>         # These must be continuous, currently a limitation.
+        >>>         "factor_1": lambda: random.uniform(0.0, 20.0),
+        >>>     })
+        >>> tune.run({...}, num_samples=8, scheduler=pb2)
+    """
+
+    def __init__(self,
+                 time_attr: str = "time_total_s",
+                 reward_attr: Optional[str] = None,
+                 metric: Optional[str] = None,
+                 mode: Optional[str] = None,
+                 perturbation_interval: float = 60.0,
+                 hyperparam_bounds: Dict = None,
+                 quantile_fraction: float = 0.25,
+                 log_config: bool = True,
+                 require_attrs: bool = True,
+                 synch: bool = False):
+
+        if GPy is None:
+            raise RuntimeError("Need to install GPy to use PB2")
+
+        resample_probability = 0
+        custom_explore_fn = None
+        hyperparam_mutations = None
+
+        super(PopulationBasedTraining, self).__init__(
+            self, time_attr, reward_attr, metric, mode, perturbation_interval,
+            hyperparam_mutations, quantile_fraction, resample_probability,
+            custom_explore_fn, log_config, require_attrs, synch)
+
+        self.last_exploration_time = 0  # when we last explored
+        self.data = pd.DataFrame()
+
+        self.hyperparam_bounds = hyperparam_bounds
+
+        # Current = trials running that have already re-started after reaching
+        #           the checkpoint. When exploring we care if these trials
+        #           are already in or scheduled to be in the next round. 
+        self.current = None 
+
+    def _save_trial_state(self, state, time, result, trial):
+
+        score = super(PopulationBasedTraining, self)._save_trial_state(
+            state, time, result, trial)
+
+        # Data logging for PB2.
+
+        # Collect hyperparams names and current values for this trial.
+        names = []
+        values = []
+        for key in self._hyperparam_mutations.keys():
```

Comment:
I think `self._hyperparam_mutations` is a dict, so we can do this (which may be a little faster):
```suggestion
        for key in self._hyperparam_mutations:
```